Add nRF9280 SiP and board

lemrey opened the pull request:
Add initial support for nRF9280 SiP and PDK.

The nRF9280 SiP is based on the multicore nRF9230 SoC.
This patch adds build targets for the application, radio and PPR processors.

lemrey committed:
dts: Add initial support for nRF9280 SiP

Add definition of the nRF9280 SiP with its Application,
Radio, and Peripheral Processor (PPR) cores and a basic set
of peripherals: GRTC, GPIOs, GPIOTE, and UARTs and few others.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>
Co-authored-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

lemrey committed:
soc: nordic: Add initial support for nRF9280 SiP

The nRF9280 is a SiP (System-in-Package) consisting of the nRF9230 SoC
and additional components such as PMIC and others. Additionally,
the nRF9230 contains several CPUs, similarly to the nRF54h20 SoC.

Update nrfx glue, and add necessary Kconfig and initialization code
to allow building for nRF9280 targets: CPU, Radio and PPR cores.

The nRF9280 is used for all user build targets and Kconfigs,
whereas the nRF9230 is used as the build target for the MDK.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>
Co-authored-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

lemrey committed:
drivers: cache: nrf: enable CACHE_NRF_PATCH_LINEADDR for nRF92

Enable CACHE_NRF_PATCH_LINEADDR for nRF92 by default,
similarly to how it's enabled for nRF54H.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey committed:
drivers: serial: nrf: disable legacy shim for nRF92

Disable the legacy shim due to DPPI support currently missing.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey committed:
scripts: west: runners: add initial nRF92 support

Add initial support flashing and erasing the application and
radio cores of the nRF9280.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey committed:
boards: nordic: add initial support for nRF9280 PDK

Add board support for Application, Radio and PPR cores
of the nRF9230 SoC / nRF9280 SiP on the nRF9280 PDK board.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>
Co-authored-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

bjarki-andreasen reviewed (commented):
Looks good :) only initial note I have is inconsistent naming of the board files, just use underscores in names instead of sometimes using dashes :) like `nrf9280-ipc_conf.dtsi` -> `nrf9280_ipc_conf.dtsi`

hubertmis commented on soc/nordic/nrf92/mpu_regions.c:
```diff
@@ -0,0 +1,40 @@
+/*
```
This file looks like it could be reused for nrf54h. Would it be possible to share it?

hubertmis commented on soc/nordic/nrf92/soc.c:
```diff
@@ -0,0 +1,114 @@
+	 *  WFI the power domain will be correctly retained.
+	 */
+
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, true);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, true);
+
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, true);
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, true);
```
It looks like both nRF54H and nRF92 will require cleanup when Suspend-to-Idle gets implemented. We'll need to create a plan around this: if we target one SoC at a time or implement PM for all simultaneously.

hubertmis commented on drivers/serial/Kconfig.nrfx:
```diff
@@ -33,6 +33,7 @@ config UART_NRFX_UARTE_LEGACY_SHIM
 	depends on UART_NRFX_UARTE
 	depends on !SOC_SERIES_NRF54LX
 	depends on RISCV || !SOC_SERIES_NRF54HX
+	depends on RISCV || !SOC_SERIES_NRF92X
```
Is this any different from oneliner?

`depends on !(SOC_SERIES_NRF54HX || SOC_SERIES_NRF92X) || RISCV`

At least I find the current two lines confusing.
The oneliner above I understand as "this is unavailable on nRF54L".
This one I proposed I read as "this is unavailable on the SoCs from the list (54H, 92) for other CPUs than RISC-Vs".

Ok, maybe by changing the order it would be easier to understand:

```
depends on !SOC_SERIES_NRF54HX || RISCV
depends on !SOC_SERIES_NRF92X || RISCV
```
meaning "unavailable on 54H unless in VPR" and "... nRF92 unless in VPR"

hubertmis commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280-ipc_conf.dtsi:
```diff
@@ -0,0 +1,51 @@
+		};
+
+		cpusec_cpurad_ipc: ipc-1-3 {
+			compatible = "zephyr,ipc-icmsg";
```
I thought cpusec uses ipc-icbmsg. Is this compatible correct?

57300 replied:
> I thought cpusec uses ipc-icbmsg.

Not currently.

hubertmis commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280-memory_map.dtsi:
```diff
@@ -0,0 +1,262 @@
+
+/ {
+	reserved-memory {
+		cpuapp_ram0x_region: memory@2f011000 {
```
cpuapp_ram0x_region and cpurad_ram0x_region were recently swapped in the memory layout design to make creating cpuapp overlay files simpler. Ref http://int-ncs-doc.nordicsemi.no/main/html/internal/sw_arch/memory_maps/nrf9230.html#general-purpose-shared-ram-ram0x

lemrey replied:
Thank you, I have updated this.

hubertmis commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280-memory_map.dtsi:
```diff
@@ -0,0 +1,262 @@
+			#size-cells = <1>;
+			ranges = <0x0 0x2f0c0000 0x5000>;
+
+			/* Define another node with ownder-id 4 to allocate
+			 * this area and share it with the modem code.
+			 */
```
This structure (reserved-memory/nordic,owned-memory/nordic,owned-memory) looks confusing.

I'm thinking about something a little different like:

```
reserved-memory {
    cpuapp_cpumod_ram0x_region: memory@2f0c0000 {
        reg = <0x2f0c0000 DT_SIZE_K(20)>;
        #address-cells = <1>;
        #size-cells = <1>;
        ranges = <0x0 0x2f0c0000 DT_SIZE_K(20)>;

        /* Memory partitions */
        cpuapp_cpumod_ipc_shm_ctrl: memory@0 {
            reg = <0x0 0xce0>;
        };
        /* TX heap, user defined */
        cpuapp_cpumod_ipc_shm_heap: memory@ce0 {
            reg = <0xce0 0x2080>;
        };
        /* RX heap, user defined */
        cpumod_cpuapp_ipc_shm_heap: memory@2d60 {
            reg = <0x2d60 0x2000>;
        };

        /* Shared memory ownership */
        cpuapp_cpumod_ipc_shm: memory@2 {
            compatible = "nordic,owned-memory";
            reg = <0x0 DT_SIZE_K(20)>;
            owner-id = <2>;
            perm-read;
            perm-write;
            status = "disabled";
        };
        cpumod_cpuapp_ipc_shm: memory@4 {
            compatible = "nordic,owned-memory";
            reg = <0x0 DT_SIZE_K(20)>;
            owner-id = <4>;
            perm-read;
            perm-write;
            status = "disabled";
        };
    };
};
```

What do you think?

hubertmis replied:
BTW, it would be perfect for this case if `owner-id` could be a list like `owner-id = <2 4>;`. But nrf-regtool doesn't support this. Or does it?

57300 replied:
> BTW, it would be perfect for this case if `owner-id` could be a list like `owner-id = <2 4>;`. But nrf-regtool doesn't support this. Or does it?

The intended solution was to use the `nordic,access` property from #72273. At first, it wasn't supposed to be merged until the relevant secure domain work was completed.

We could start using it now though, but only for the multi-owner use case, and without deprecating the `perm-*` properties yet. I can update the PR if that sounds reasonable.

hubertmis replied:
Right, I forgot about `nordic,access`. That's the correct solution. For me, it's acceptable to keep (reserved-memory/nordic,owned-memory/nordic,owned-memory) temporarily. Just add a TODO comment to refactor it once #72273 is merged

lemrey replied:
I went with your suggestion here, and left a TODO.

hubertmis commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280_cpuapp.dts:
```diff
@@ -0,0 +1,308 @@
+	status = "okay";
+	interrupts = <96 NRF_DEFAULT_IRQ_PRIORITY>;
+	interrupt-names = "irq0";
+	/* irq0: 0: cpuapp-cpusec, 6: cpuapp-cpusys, 13: cpuapp-cpuppr, 18: cpuapp-cpurad
```
I would comment "cpusec->cpuapp, 6: cpusys->cpuapp, ..." to indicate the signal direction

57300 commented on dts/arm/nordic/nrf9280_cpuapp.dtsi:
```diff
@@ -0,0 +1,65 @@
+/*
+ * TODO
+ * &gpiote131 {
+ *	interrupts = <106 NRF_DEFAULT_IRQ_PRIORITY>;
```
This can be 107 for APP/RAD and 106 for PPR.

57300 commented on dts/common/nordic/nrf9280.dtsi:
```diff
@@ -0,0 +1,1207 @@
+					     <15 NRF_DEFAULT_IRQ_PRIORITY>;
+				#mbox-cells = <1>;
+				nordic,tasks = <16>;
+				nordic,tasks-mask = <0xfffffff0>;
```
```suggestion
				nordic,tasks-mask = <0x0000fff0>;
```

lemrey replied:
🤔 is this intended to be different from the configuration on the nRF54h20? Just double checking.

57300 replied:
This value comes from IPS, where it's different from nRF54H20. However, since the task count is 16, they seem pretty much equivalent here.

57300 commented on dts/common/nordic/nrf9280.dtsi:
```diff
@@ -0,0 +1,1207 @@
+			/* TODO check */
+			can121: can@8db000 {
+				compatible = "nordic,nrf-can";
+				reg = <0x8db000 0x400>, <0x2fbef800 0x800>, <0x2fbe8000 0x7800>;
```
```suggestion
				reg = <0x8db000 0x400>, <0x2fbf7800 0x800>, <0x2fbf0000 0x7800>;
```

57300 commented on dts/common/nordic/nrf9280.dtsi:
```diff
@@ -0,0 +1,1207 @@
+				compatible = "nordic,nrf-can";
+				reg = <0x8db000 0x400>, <0x2fbef800 0x800>, <0x2fbe8000 0x7800>;
+				reg-names = "wrapper", "m_can", "message_ram";
+				interrupts = <216 NRF_DEFAULT_IRQ_PRIORITY>;
```
```suggestion
				interrupts = <219 NRF_DEFAULT_IRQ_PRIORITY>;
```

57300 commented on dts/common/nordic/nrf9280.dtsi:
```diff
@@ -0,0 +1,1207 @@
+					status = "disabled";
+					#mbox-cells = <1>;
+					nordic,tasks = <16>;
+					nordic,tasks-mask = <0xfffffff0>;
```
```suggestion
					nordic,tasks-mask = <0x0000fff0>;
```

57300 commented on dts/common/nordic/nrf9280.dtsi:
```diff
@@ -0,0 +1,1207 @@
+				#gpio-cells = <2>;
+				gpio-controller;
+				gpiote-instance = <&gpiote130>;
+				ngpios = <12>;
```
```suggestion
				ngpios = <13>;
```

57300 commented on include/zephyr/dt-bindings/misc/nordic-domain-id-nrf9230.h:
```diff
@@ -0,0 +1,14 @@
+
+#define NRF_DOMAIN_ID_APPLICATION 2
+#define NRF_DOMAIN_ID_RADIOCORE   3
+#define NRF_DOMAIN_ID_CELL        4
```
Rename to match the spec.
```suggestion
#define NRF_DOMAIN_ID_CELLCORE    4
```

57300 commented on include/zephyr/dt-bindings/misc/nordic-owner-id-nrf9230.h:
```diff
@@ -0,0 +1,13 @@
+#define NRF_OWNER_ID_NONE        0
+#define NRF_OWNER_ID_APPLICATION 2
+#define NRF_OWNER_ID_RADIOCORE   3
+#define NRF_OWNER_ID_CELLCORE    4
```
Rename to match the spec.
```suggestion
#define NRF_OWNER_ID_CELL        4
```

57300 commented on scripts/west_commands/runners/nrf_common.py:
```diff
@@ -254,9 +256,12 @@ def program_hex(self):
+        if self.family in ('NRF54H_FAMILY', 'NRF92_FAMILY'):
             erase_arg = 'ERASE_NONE'
 
+            cpuapp = self.build_conf.getboolean('CONFIG_SOC_NRF54H20_CPUAPP') or self.build_conf.getboolean('CONFIG_SOC_NRF9280_CPUAPP')
+            cpurad = self.build_conf.getboolean('CONFIG_SOC_NRF54H20_CPURAD') or self.build_conf.getboolean('CONFIG_SOC_NRF9280_CPURAD')
```
We could use more generic conditions here. How about checking the board qualifiers for `/cpuapp` etc?

57300 replied:
Fair point. We can try to improve this in future PRs.

57300 commented on soc/nordic/nrf92/mpu_regions.c:
```diff
@@ -0,0 +1,40 @@
+#define USBHS_SIZE	DT_REG_SIZE_BY_NAME(DT_NODELABEL(usbhs), core)
+
+#define CAN120_BASE	DT_REG_ADDR_BY_NAME(DT_NODELABEL(can120), message_ram)
+#define CAN120_SIZE	DT_REG_SIZE_BY_NAME(DT_NODELABEL(can120), message_ram) + \
```
I guess CAN121 should also be accounted for here.

57300 commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280_cpurad.dts:
```diff
@@ -0,0 +1,120 @@
+		zephyr,ieee802154 = &cpurad_ieee802154;
+		zephyr,bt-hci-ipc = &ipc0;
+		nordic,802154-spinel-ipc = &ipc0;
+		zephyr,entropy = &prng;
+	};
+	prng: prng {
+		compatible = "nordic,entropy-prng";
+		status = "okay";
+	};
```
Do we want this upstream? The binding is only downstream.

57300 reviewed (changes requested):
Looks good so far. Some initial observations.

57300 commented on dts/common/nordic/nrf9280.dtsi:
```diff
@@ -0,0 +1,1207 @@
+					nordic,tasks-mask = <0xfffffff0>;
+				};
+
+				cpuppr_clic: interrupt-controller@1000 {
+					compatible = "nordic,nrf-clic";
+					reg = <0x1000 0x3000>;
+					status = "disabled";
+					#interrupt-cells = <2>;
+					interrupt-controller;
+					#address-cells = <1>;
+				};
```
> I don't think 0x909000 is the correct CLIC address. Anyway, this peripheral is internal VPR peripheral so probably should be removed from this common file.

nrf-regtool depends on CLIC nodes being visible from any core. We use that for interrupt mapping between cores.

Also, judging by the address of this node on nRF9230 EngB (0x5F909000 is correct), it appears to be accessible to PPR's owner, so I think it can stay here for now.

(Note: on other chips, a CLIC may be private to a given VPR and inaccessible to all other cores. This can still be expressed using the `ranges` property - or lack thereof - on the parent bus, but the CLIC node itself would have to be moved.)

hubertmis replied:
> Also, judging by the address of this node on nRF9230 EngB (0x5F909000 is correct), it appears to be in the global domain and accessible to PPR's owner, so I think it can stay here for now.

Can you point me to the IPS chapter describing this address? I could not find it. That's why I suggested removing it.

hubertmis replied:
Thanks for clarifying. The address is correct and my comment is irrelevant.

57300 commented on dts/common/nordic/nrf9280.dtsi:
```diff
@@ -0,0 +1,1207 @@
+		};
+	};
+
+	cpuapp_ppb: cpuapp-ppb-bus {
```
> Why this node is present in a common file instead of in a cpuapp specific file? I don't think any other CPU should have it included in their device trees.

The private bus nodes were introduced to help keep the entire SoC description in one file.

hubertmis replied:
What is the point of the entire SoC description in one file? I don't think any other CPU is interested in App having something at its PPB.

Private peripherals' nodes in the device tree files are unlike the common memory description. The value of the common memory map makes it easier to avoid mistakes when repartitioning the system (replacement of missing system device tree).

57300 replied:
The `<soc>.dtsi` is included by every CPU, but anything that shouldn't be visible is masked out using `/delete-node/`. Think of each `<soc>_<cpu>.dtsi` as a projection of `<soc>.dtsi`, which mostly subtracts content and doesn't add too much new stuff. The point is to organize a relatively complex SoC as tightly as possible, in as few files as possible. It's also inspired by would-be operations on a system devicetree.

hubertmis replied:
It's confusing for me to create cpu-specific nodes in common files only to `delete-node` them later in cpu-specific files. (but at least I see the reason for these long `/delete-node/` lists I've seen recently so many times)

The system device tree would be another story, but I guess it would generate the CPU-specific device tree files containing only nodes related to the target CPU. So I imagined we aim to mimic the result of the system device tree tooling.

It's not a strong opinion. I see your point. The fact that it makes reading the files harder for me does not mean the result is the same for everyone.

57300 replied:
> The fact that it makes reading the files harder for me does not mean the result is the same for everyone.

I can see that it can be difficult to get used to. This approach has only existed in-tree for a few months, so it's good to re-examine it every once in a while :)

hubertmis replied:
Ok, let's keep it unified for now and we can open discussion to change it globally for all Nordic .dts files in another PR

lemrey committed:
soc: nordic: move mpu_regions.c to common folder and rename

Move mpu_region.c to common folder, to re-use with nRF92.
Rename it to nrf54hx_nrf92x_mpu_regions.c to indicate
which product series it applies to.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey committed:
soc: nordic: common: add CAN121 to nrf54hx_nrf92x_mpu_regions.c

Add support for CAN121, if present in DT.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey commented on soc/nordic/nrf92/Kconfig.defconfig:
```diff
@@ -0,0 +1,42 @@
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 1000000 if NRF_GRTC_TIMER
+	default 32768 if NRF_RTC_TIMER
```
Thanks, I have removed this.

lemrey reviewed (commented):
Have addressed the comments.
Also, made the following changes:
- `cpumod` -> `cpucell` rename
- doubled the application RAM (`256`->`512`)
- doubled TX/RX heaps towards cellcore `8` -> `16`
- aligned the IPC buffers towards cellcore to 4kb

lemrey commented on scripts/west_commands/runners/nrfutil.py:
```diff
@@ -61,7 +61,8 @@ def _exec(self, args):
-                            pld['data']['progress']['progressPercentage'] == 0
+                            # temporary fix for nrutil device
+                            pld['data']['progress']['progress_percentage'] == 0
                         ):
```
What's currently upstream is correct, so I have reverted this change.

lemrey commented on scripts/west_commands/runners/nrfutil.py:
```diff
@@ -61,7 +61,7 @@ def _exec(self, args):
                         if (
                             pld['type'] == 'task_progress' and
-                            # temporary fix for nrutil device
+                            # temporary fix for nrfutil
```
Thanks, I have fixed this.

lemrey commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280_cpuapp_defconfig:
```diff
@@ -0,0 +1,28 @@
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+
+CONFIG_USE_DT_CODE_PARTITION=y
```
Perhaps. This is how it's done currently for nRF54h20, maybe @anangl can comment.

lemrey commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280_cpuapp_defconfig:
```diff
@@ -0,0 +1,28 @@
+# Enable hardware stack protection
+CONFIG_HW_STACK_PROTECTION=y
+
+# MPU-based null-pointer dereferencing detection cannot be applied
+# as the (0x0 - 0x400) region is unmapped for this target.
+CONFIG_NULL_POINTER_EXCEPTION_DETECTION_NONE=y
```
This one too, it's defined here for nRF54h20..

lemrey commented on dts/common/nordic/nrf9280.dtsi:
```diff
@@ -0,0 +1,1207 @@
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		/*
+		 * TODO, check. disable for now
+		suit_storage_partition: memory@e1eb000 {
+			reg = <0xe1eb000 DT_SIZE_K(24)>;
+		};
+		*/
```
Hey @hubertmis , I copied this from the nRF54h20. I think the placement is fine, but could you please double check?

hubertmis replied:
@tomchy, please have a look. I don't know SUIT requirements for nRF92

lemrey replied:
According to @tomchy this is not supposed to work on nRF92 yet, so I have removed this.

nordicjm commented on soc/nordic/nrf92/Kconfig.soc:
```diff
@@ -0,0 +1,62 @@
+# The SiP selects the actual SoC complete with engineer revision and appropriate CPU
+config SOC_NRF9280
+	bool
+	select SOC_NRF9230_ENGB
+	select SOC_NRF9230_ENGB_CPUAPP if SOC_NRF9280_CPUAPP
+	select SOC_NRF9230_ENGB_CPURAD if SOC_NRF9280_CPURAD
+	select SOC_NRF9230_ENGB_CPUPPR if SOC_NRF9280_CPUPPR
```
they need to select it in Kconfig.soc, not Kconfig

lemrey replied:
Thanks, that seemed to work

lemrey committed:
modules: hal_nordic: nrf: add nRF9280 support

Enable NRFS support for nRF9280, including DVFS.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey committed:
tests: lib: cpp: cxx: add nRF9280 to exclusion list

Nordic HAL is not compatible with C++98.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey committed:
tests: drivers: watchdog: wdt_error_cases: fix compilation for nRF9280

Fix compilation by defining the watchdog flags for nRF9280 as well.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>
Co-authored-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

lemrey committed:
tests: drivers: watchdog: wdt_basic_api: fix compilation for nRF9280

Fix compilation by providing an overlay file for cpuapp.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>
Co-authored-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

lemrey committed:
samples: drivers: watchdog: add overlay file for nRF9280 cpuapp

Add an overlay to enable compilation.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey committed:
tests: arch: arm: arm_irq_vector_table: fix compilation for nRF9280

Fix compilation for nRF9280.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

hubertmis commented on soc/nordic/common/CMakeLists.txt:
```diff
@@ -12,6 +12,7 @@ if(CONFIG_ARM)
 endif()
 
 zephyr_library_sources_ifdef(CONFIG_POWEROFF poweroff.c)
+zephyr_library_sources_ifdef(CONFIG_CPU_HAS_CUSTOM_FIXED_SOC_MPU_REGIONS mpu_regions.c)
```
This seems applicable only to selected SoCs. How about renaming the file to something like nrf54h_nrf92_mpu_regions.c and adding some more configuration guards to make sure it's not included in nRF54L or nRF91 builds?

hubertmis replied:
When we resolve this comment, I'll approve.

lemrey replied:
Alright, I have renamed it to `nrf54hx_nrf92x_mpu_regions.c` and update the `if` condition to check for the appropriate SoC series.

hubertmis commented on soc/nordic/common/mpu_regions.c:
```diff
@@ -32,6 +36,10 @@ static struct arm_mpu_region mpu_regions[] = {
 			 REGION_RAM_NOCACHE_ATTR(CAN120_BASE, CAN120_SIZE)),
 #endif
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(can121), okay)
+	MPU_REGION_ENTRY("CAN121_MCAN", CAN121_BASE,
```
I can't find the MCAN addresses in IPS, but from the bus configuration, it seems possible to cover both 120 and 121 instances with a single MPU entry spanning both. What do you think?

57300 replied:
It is indeed possible to combine them, but I think it's OK to do this later.

hubertmis commented on soc/nordic/nrf92/Kconfig:
```diff
@@ -0,0 +1,49 @@
+	select NRFS_HAS_DVFS_SERVICE
+	select NRFS_HAS_MRAM_SERVICE
+	select NRFS_HAS_TEMP_SERVICE
+	select NRFS_HAS_VBUS_DETECTOR_SERVICE
```
It's the same for nRF54H. It creates the limitation that USB cannot be properly used from the Radio CPU. We'll fix it in the future. For now we can proceed with the limitation.

hubertmis commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280_cpurad.dts:
```diff
@@ -0,0 +1,115 @@
+	status = "okay";
+	interrupts = <96 NRF_DEFAULT_IRQ_PRIORITY>;
+	interrupt-names = "irq0";
+	/* irq0: 0: cpurad-cpusec, 6: cpurad-cpusys, 12: cpurad-cpuapp */
```
Like in cpuapp, I would change the comment to cpusec->cpurad, ... to indicate the signal direction.

lemrey replied:
Thank you, I have updated this and also rephrased it to make it more explicit.

anhmolt committed:
drivers: adc: nrf: Update adc driver to support nRF9280

Expands driver to support nRF9280 similar to nRF54H20 device.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

anhmolt committed:
tests: drivers: adc: fix compilation for nRF9280

Fix compilation by providing an overlay file for cpuapp.

Signed-off-by: Andreas Moltumyr <andreas.moltumyr@nordicsemi.no>

57300 commented on dts/common/nordic/nrf9280.dtsi:
```diff
@@ -0,0 +1,1270 @@
+				compatible = "nordic,nrf-uarte";
+				reg = <0x8e6000 0x1000>;
+				status = "disabled";
+				interrupts = <229 NRF_DEFAULT_IRQ_PRIORITY>;
```
```suggestion
				interrupts = <230 NRF_DEFAULT_IRQ_PRIORITY>;
```

lemrey replied:
Updated

57300 commented on soc/nordic/common/CMakeLists.txt:
```diff
@@ -13,6 +13,10 @@ endif()
 zephyr_library_sources_ifdef(CONFIG_POWEROFF poweroff.c)
 
+if((CONFIG_SOC_SERIES_NRF54HX OR CONFIG_SOC_SERIES_NRF92X) AND CONFIG_CPU_HAS_CUSTOM_FIXED_SOC_MPU_REGIONS)
+zephyr_library_sources(nrf54hx_nrf92x_mpu_regions.c)
```
nit: Missing indent
```suggestion
  zephyr_library_sources(nrf54hx_nrf92x_mpu_regions.c)
```

lemrey replied:
Fixed

57300 commented on boards/nordic/nrf9280pdk/Kconfig.nrf9280pdk:
```diff
@@ -0,0 +1,8 @@
+	select SOC_NRF9280_CPUAPP if BOARD_NRF9280PDK_NRF9280_CPUAPP
+	select SOC_NRF9280_CPURAD if BOARD_NRF9280PDK_NRF9280_CPURAD
+	select SOC_NRF9280_CPUPPR if BOARD_NRF9280PDK_NRF9280_CPUPPR || \
+				      BOARD_NRF9280PDK_NRF9280_CPUPPR_XIP
```
nit: Whitespace
```suggestion
				     BOARD_NRF9280PDK_NRF9280_CPUPPR_XIP
```

lemrey replied:
Updated

57300 commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280-memory_map.dtsi:
```diff
@@ -0,0 +1,271 @@
+
+		/* TODO:
+		 * refactor after https://github.com/zephyrproject-rtos/zephyr/pull/72273
+		 * to use the `access` binding to specify shared memory ownership.
```
I suggest attaching this comment to `cpucell_cpuapp_ipc_shm`. That's the node we will be able to drop later, once we can set multiple owners for the `cpuapp_cpucell_ram0x_region` right below.

lemrey replied:
Ok, moved

57300 commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280-memory_map.dtsi:
```diff
@@ -0,0 +1,271 @@
+				reg = <0x1000 0x4000>;
+			};
+			/* RX heap, user defined */
+			cpucell_cpuapp_ipc_shm_heap: memory@5000 {
```
nit: For consistency, there should be newlines between nodes.

lemrey replied:
Added

57300 commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280-memory_map.dtsi:
```diff
@@ -0,0 +1,271 @@
+		};
+
+		/* Shared memory ownership */
+		cpuapp_cpucell_ipc_shm: memory@2 {
```
This node seems redundant. Can we drop it and use `cpuapp_cpucell_ram0x_region` instead?

lemrey replied:
Yes, it's a bit artificial but I suggest keeping it as-is because it keeps the whole thing clearer.

57300 commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280-memory_map.dtsi:
```diff
@@ -0,0 +1,271 @@
+			perm-write;
+			status = "disabled";
+		};
+		cpucell_cpuapp_ipc_shm: memory@4 {
```
The unit address is wrong; it needs to match the value of `reg`. Be careful though: we can't have two nodes with the same name - `memory@2f0d0000` - so this one must be different somehow.
```suggestion
		cpucell_cpuapp_ipc_shm: memory-alt@2f0d0000 {
```

lemrey replied:
Agreed offline to keep as-is, given that it does not actually generate proper warnings (just messages), unlike having two nodes with the same address.

57300 replied:
For reference, the messages that currently get printed:
```
unit address and first address in 'reg' (0x2f0d0000) don't match for /reserved-memory/memory@2
unit address and first address in 'reg' (0x2f0d0000) don't match for /reserved-memory/memory@4
```
are edtlib warnings, but they should be harmless. They were meant to replicate the `simple_bus_reg` warning from dtc, but they do so incorrectly, which is why the actual dtc warning doesn't show. See #57059

57300 commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -62,9 +62,9 @@ BUILD_ASSERT((NRF_SAADC_AIN0 == NRF_SAADC_INPUT_AIN0) &&
+#if defined(CONFIG_SOC_NRF54H20) || defined(CONFIG_SOC_NRF9280)
 
-/* nRF54H20 always uses bounce buffers in RAM */
+/* nRF54H20 and nrf9280 always use bounce buffers in RAM */
```
nit:
```suggestion
/* nRF54H20 and nRF9280 always use bounce buffers in RAM */
```

lemrey replied:
Updated

57300 commented on tests/arch/arm/arm_irq_vector_table/src/arm_irq_vector_table.c:
```diff
@@ -27,8 +27,8 @@
-#elif defined(CONFIG_SOC_SERIES_NRF54HX)
-/* For nRF54H Series, use BELLBOARD_0-2 interrupt lines. */
+#elif defined(CONFIG_SOC_SERIES_NRF54HX) || defined(CONFIG_SOC_SERIES_NRF92X)
+/* For nRF54H and NRF92 Series, use BELLBOARD_0-2 interrupt lines. */
```
nit:
```suggestion
/* For nRF54H and nRF92 Series, use BELLBOARD_0-2 interrupt lines. */
```

lemrey replied:
Updated

57300 commented on tests/arch/arm/arm_irq_vector_table/src/arm_irq_vector_table.c:
```diff
@@ -154,7 +154,8 @@ void nrfx_power_clock_irq_handler(void);
 #define TIMER_IRQ_HANDLER	timer0_nrf_isr
 #define TIMER_IRQ_NUM		TIMER0_IRQn
-#elif defined(CONFIG_SOC_SERIES_NRF54LX) || defined(CONFIG_SOC_SERIES_NRF54HX)
+#elif defined(CONFIG_SOC_SERIES_NRF54LX) || defined(CONFIG_SOC_SERIES_NRF54HX) \
+	|| defined(CONFIG_SOC_SERIES_NRF92X)
```
nit: Misplaced or
```suggestion
#elif defined(CONFIG_SOC_SERIES_NRF54LX) || defined(CONFIG_SOC_SERIES_NRF54HX) || \
	defined(CONFIG_SOC_SERIES_NRF92X)
```

lemrey replied:
Moved

57300 commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280-memory_map.dtsi:
```diff
@@ -0,0 +1,271 @@
+
+		cpuppr_ram3x_region: memory@2fc00000 {
+			compatible = "nordic,owned-memory";
+			reg = <0x2fc00000 DT_SIZE_K(16)>;
```
For PPR code, 16 KB is not a lot. Should we increase it to at least 24?

lemrey replied:
ok, increased

57300 commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280_cpuppr.yaml:
```diff
@@ -0,0 +1,18 @@
+toolchain:
+  - zephyr
+sysbuild: true
+ram: 16
```
This should match the size of `cpuppr_code_data`
```suggestion
ram: 14
```

lemrey replied:
Updated to match the size of the updated `cpuppr_code_data` region.

57300 commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280_cpuppr.yaml:
```diff
@@ -0,0 +1,18 @@
+  - zephyr
+sysbuild: true
+ram: 16
+flash: 64
```
I believe `flash` should equal `ram` in the non-XIP case.

lemrey replied:
Updated

57300 commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280_cpuppr_xip.yaml:
```diff
@@ -0,0 +1,14 @@
+toolchain:
+  - zephyr
+sysbuild: true
+ram: 62
```
This should match the size of `cpuppr_code_data`
```suggestion
ram: 14
```

lemrey replied:
Updated to match the size of the updated `cpuppr_code_data` region.

57300 commented on boards/nordic/nrf9280pdk/nrf9280pdk_nrf9280_cpurad.yaml:
```diff
@@ -0,0 +1,19 @@
+  - zephyr
+sysbuild: true
+ram: 32
+flash: 368
```
This should match the size of `cpurad_slot0_partition`
```suggestion
flash: 256
```

lemrey replied:
Updated
